### PR TITLE
Add trackvis ushape

### PR DESCRIPTION
**Pull request by arnaudbore**
Dummy version of the extract ushape. I guess can be optimized wait for suggestion.

### CODE REVIEW

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-375/5/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-375/6/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-375/7/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-375/8/

File: scripts/scil_extract_ushape.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+
+def _build_arg_parser():
+    p = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter,
+                                description=__doc__)
+    p.add_argument('in_tractogram',
+                   help='Tractogram input file name.')
+    p.add_argument('out_tractogram',
+                   help='Output tractogram with ushape.')
+    p.add_argument('--ufactor', default=[0.5, 1], type=float, nargs=2,
```
Comment (frheault) [inline scripts/scil_extract_ushape.py:48]: Could you add a second line (or in the docstring) to explain what are those two?
What if I put 0,1 or 0,0 or 1,1 or 0.5, 0.5?
Practically, what are they for?

File: scripts/scil_extract_ushape.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    parser = _build_arg_parser()
+    args = parser.parse_args()
+
+    assert_inputs_exist(parser, args.in_tractogram)
+    assert_outputs_exist(parser, args, args.out_tractogram,
+                         optional=args.remaining_tractogram)
+    check_tracts_same_format(parser, [args.in_tractogram, args.out_tractogram,
+                                      args.remaining_tractogram])
+
+    #if min(args.ufactor) <= 0 or max(args.ufactor) > 1:
```
Comment (frheault) [inline scripts/scil_extract_ushape.py:72]: This should stay, right?

File: scripts/scil_extract_ushape.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    #    parser.error('U-factor "{}" '.format(args.ufactor) +
+    #                 'must be greater than 0 and lower or equal to 1.')
+
+    tractogram = load_tractogram_with_reference(
+        parser, args, args.in_tractogram)
+
+    streamlines = tractogram.streamlines
+
+    ids_c = []
+
```
Comment (frheault) [inline scripts/scil_extract_ushape.py:82]: Why so much spacing?

File: scripts/scil_extract_ushape.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        parser.error(
+            'Zero or one streamline in {}'.format(args.in_tractogram) +
+            '. The file must have more than one streamline.')
+
+    if len(ids_c) > 0:
+        sft_c = filter_tractogram_data(tractogram, ids_c)
+        save_tractogram(sft_c, args.out_tractogram)
+    else:
+        logging.warning(
+            'No clean streamlines in {}'.format(args.in_tractogram))
```
Comment (frheault) [inline scripts/scil_extract_ushape.py:98]: Why "clean"? U Shape or not, its about cleanliness, right?
No U-Shape maybe?

File: scilpy/tractanalysis/features.py [hunk 5]
```diff
@@ -10,6 +10,45 @@
 import numpy as np
 
 
+def detect_ushape(streamlines, ufactor=[0.5, 1]):
```
Comment (frheault) [inline scilpy/tractanalysis/features.py:13]: In the function I would give a detailled example of what u factor means.
Like in the fibercub there is a u bundle, could you extract it with this function?
The AF is a BIG u shape, can it be detected?

(Also the whole docstrinc is a copy paste and there is a bla bla bla)

File: scilpy/tractanalysis/features.py [hunk 6]
```diff
@@ -45,6 +84,8 @@ def remove_loops_and_sharp_turns(streamlines,
         if tm.winding(s) < max_angle:
             ids.append(i)
             streamlines_clean.append(s)
+        else:
```
Comment (frheault) [inline scilpy/tractanalysis/features.py:87]: I don't understand this else. This wont work (adding an empty single point streamlines).
Why is it there?

File: scripts/scil_extract_ushape.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    assert_outputs_exist(parser, args, args.out_tractogram,
+                         optional=args.remaining_tractogram)
+    check_tracts_same_format(parser, [args.in_tractogram, args.out_tractogram,
+                                      args.remaining_tractogram])
+
+    #if min(args.ufactor) <= 0 or max(args.ufactor) > 1:
+    #    parser.error('U-factor "{}" '.format(args.ufactor) +
+    #                 'must be greater than 0 and lower or equal to 1.')
+
+    tractogram = load_tractogram_with_reference(
```
Comment (frheault) [inline scripts/scil_extract_ushape.py:76]: I prefer calling this `sft` (optional)

File: scripts/scil_extract_ushape.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    if len(streamlines) > 1:
+        ids_c = detect_ushape(streamlines, args.ufactor)
+        ids_l = np.setdiff1d(np.arange(len(streamlines)), ids_c)
+    else:
+        parser.error(
+            'Zero or one streamline in {}'.format(args.in_tractogram) +
+            '. The file must have more than one streamline.')
+
+    if len(ids_c) > 0:
+        sft_c = filter_tractogram_data(tractogram, ids_c)
```
Comment (frheault) [inline scripts/scil_extract_ushape.py:94]: `tractogram[ids_c]` should be enough

File: scripts/scil_extract_ushape.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+    if args.display_counts:
+        sc_bf = len(tractogram.streamlines)
+        sc_af = len(sft_c.streamlines)
+        print(json.dumps({'streamline_count_before_filtering': int(sc_bf),
+                         'streamline_count_after_filtering': int(sc_af)},
+                         indent=args.indent))
+
+    if len(ids_l) == 0:
+        logging.warning('No loops in {}'.format(args.in_tractogram))
```
Comment (frheault) [inline scripts/scil_extract_ushape.py:108]: loops?

File: scripts/scil_extract_ushape.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    elif args.remaining_tractogram:
+        sft_l = filter_tractogram_data(tractogram, ids_l)
```
Comment (frheault) [inline scripts/scil_extract_ushape.py:110]: same as above, `tractogram[ids_l]`

File: scripts/tests/test_extract_ushape.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import os
+import tempfile
+
+from scilpy.io.fetcher import get_testing_files_dict, fetch_data, get_home
+
+
+# If they already exist, this only takes 5 seconds (check md5sum)
```
Comment (frheault) [inline scripts/tests/test_extract_ushape.py:10]: why comment?

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-375/9/

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_extract_ushape.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+v3: pt3 -> pt4
+
+ufactor:dot product of  v1 X v2 and v2 X v3.
+X is the cross product of two vectors.
+
+When ufactor is close to:
+*  0 it defines straight streamlines
+*  1 it defines U-fibers
+* -1 it defines S-fibers
+----------------------------------------------------------------------------
```
Comment (frheault) [inline scripts/scil_extract_ushape.py:25]: Why the line?

File: scripts/scil_extract_ushape.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+    if args.display_counts:
+        sc_bf = len(sft.streamlines)
+        sc_af = len(sft_c.streamlines)
+        print(json.dumps({'streamline_count_before_filtering': int(sc_bf),
+                         'streamline_count_after_filtering': int(sc_af)},
+                         indent=args.indent))
+
+    if len(ids_l) == 0:
+        logging.warning('No loops in {}'.format(args.in_tractogram))
```
Comment (frheault) [inline scripts/scil_extract_ushape.py:115]: Why loops?

File: scripts/scil_extract_ushape.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    elif args.remaining_tractogram:
+        sft_l = sft[ids_l]
+        save_tractogram(sft_l, args.remaining_tractogram)
```
Comment (frheault) [inline scripts/scil_extract_ushape.py:118]: Could be `save_tractogram(sft[ids_l], args.remaining_tractogram)` (same above)

**Review by frheault — APPROVED** [review]

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-375/14/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-375/15/

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Pull request merged**